File: lab3-primit/java/OSTypeEnum.java
import java.util.Scanner;
import java.util.InputMismatchException;
import java.util.EnumSet;

public class OSTypeEnum
{
    enum OS { ANDROID, IOS, MACOSX, WINDOWS8, WP8, VXWORKS }
    
    enum OSType { DESKTOP, EMBEDDED, MOBILE }
    
    private static <E extends Enum<E>> E getEnumElement(String elementTypeName, Class<E> elementType)
    {
        boolean haveResult = false;
        E result = null;
        Scanner stdin = new Scanner(System.in);
        
        while ( ! haveResult )
        {
            System.out.print("Input " + elementTypeName + ": ");
            try
            {
                result = Enum.valueOf(elementType, stdin.next().toUpperCase());
                haveResult = true;
            }
            catch (IllegalArgumentException e)
            {
                System.out.println("Not a valid " + elementTypeName + ".");
                stdin.nextLine(); // skip the invalid input
            }
        }
        
        return result;
    }
  
    private static OSType os2OSType(OS os)
    {
        OSType type = null;
        
        switch (os)
        {
        case ANDROID:
        case IOS:
        case WP8:
            type = OSType.MOBILE;
            break;
        case WINDOWS8:
        case MACOSX:
            type = OSType.DESKTOP;
            break;
        case VXWORKS:
            type = OSType.EMBEDDED;
            break;
        }
        
        return type;
    }

    private static OS ost2OS(OSType ost) 
    {
        OS os = null;
        switch (ost) 
        {
            case MOBILE:
                os = OS.ANDROID;
                break;
            case DESKTOP:
                os = OS.WINDOWS8;
                break;
            case EMBEDDED:
                os = OS.VXWORKS;
                break;
        }

        return os;
    }

    private static void showOsToType() {
        System.out.print("Known OSs = ");
        for (OS t : EnumSet.allOf(OS.class)) 
        {
            System.out.print(t + " ");
        }
        System.out.println();
        
        OS os = getEnumElement("operating system", OS.class);
        System.out.println(os + " is of type: " + os2OSType(os));
    }

    private static void showTypeToOs() 
    {
        System.out.print("Known Types = ");
        for (OSType t : EnumSet.allOf(OSType.class)) 
        {
            System.out.print(t + " ");
        }
        System.out.println();
        
        OSType ost = getEnumElement("operating system type", OSType.class);
        System.out.println(ost2OS(ost) + " is an example of a " + ost + " operating system");
    }

    public static void main(String[] args)
    {
        showTypeToOs();
    }
}

File: lab3-primit/haskell/ostypeEnum.hs
module Main where

import System.IO (hSetBuffering, stdout, BufferMode(..))
import Data.Char (toUpper)

main =
    do
    initialiseIO
    getType

getOs = 
    do
    putStrLn $ "known OSs = " ++ show allOSs
    os <- getElement "operating system"
    putStrLn $ show os ++ " is of type: " ++ show (os2OSType os)

getType = 
    do
    putStrLn $ "Known Types = " ++ show allTypes
    ost <- getElement "operating system type"
    putStrLn $ show (ost2OS ost) ++ " is an example of a " ++ show ost ++ " OS"

initialiseIO =
    do
    hSetBuffering stdout NoBuffering
        -- ensure any console output is shown asap

data OS = ANDROID | IOS | MACOSX | WINDOWS8 | WP8 | VXWORKS
    deriving (Show, -- default formatting
              Read, -- default parsing
              Eq,   -- default equality testing
              Bounded, -- default minBound and maxBound
              Enum) -- default sequencing (needed for .. ranges)
data OSType = DESKTOP | EMBEDDED | MOBILE
    deriving (Show, -- default formatting
              Read, -- default parsing
              Eq,   -- default equality testing
              Bounded, -- default minBound and maxBound
              Enum) -- default sequencing (needed for .. ranges)

allOSs :: [OS] -- ie it is a list of OS elements
allOSs = [minBound .. maxBound]
allTypes :: [OSType]
allTypes = [minBound .. maxBound]

os2OSType ANDROID = MOBILE
os2OSType IOS = MOBILE
os2OSType MACOSX = DESKTOP
os2OSType WINDOWS8 = DESKTOP
os2OSType WP8 = MOBILE
os2OSType VXWORKS = EMBEDDED

ost2OS MOBILE = ANDROID
ost2OS DESKTOP = MACOSX
ost2OS EMBEDDED = VXWORKS

getElement elementTypeName =
    keepTrying
    where
    keepTrying =
        do
        showPrompt
        line <- getLine -- get whatever user types as a string
        let lineUpperCase = map toUpper line -- convert each character to upper case
        case reads lineUpperCase of -- attempt to parse the string
                -- after converting all characters to upper case
            [] -> -- no valid interpretation of the line as an element
                do
                reportError
                keepTrying -- try again - using recursion
            ((e,_) : _) -> -- at least one interpretation - it is given the name e
                    -- underscore here means "anything - its value does not matter"
                return e -- type of e is derived from context
                        -- where getElement is used
    showPrompt =
        putStr ("Input one " ++ elementTypeName ++ ": ")
    reportError =
        putStrLn ("Invalid " ++ elementTypeName ++ ", try again")
